File: Frontend/trackr-administrador/src/app/shared/http/catalogo/titulo-academico.service.ts
import { Injectable } from '@angular/core';
import { HttpClient } from '@angular/common/http';
import { Observable } from 'rxjs';
import { Estado } from '@models/catalogo/estado';
import { TituloAcademico } from '@models/catalogo/titulo-academico';

@Injectable()
export class TituloAcademicoService {
  private dataUrl = 'tituloAcademico/';

  constructor(public http: HttpClient) {}

  public consultarTodosParaSelector(): Observable<TituloAcademico[]> {
    return this.http.get<TituloAcademico[]>(this.dataUrl + `consultarTodosParaSelector`);
  }

  public consultarTodosParaGrid() {
    return this.http.get<TituloAcademico[]>(this.dataUrl + 'consultarTodosParaGrid');
  }

  public consultar(idTituloAcademico: number) {
    return this.http.get<TituloAcademico>(this.dataUrl + `consultar/${idTituloAcademico}`);
  }

  public agregar(tituloAcademico: TituloAcademico) {
    return this.http.post<void>(this.dataUrl + 'agregar', tituloAcademico);
  }

  public editar(tituloAcademico: TituloAcademico) {
    return this.http.put<void>(this.dataUrl + 'editar', tituloAcademico);
  }

  public eliminar(idTituloAcademico: number) {
    return this.http.delete<void>(this.dataUrl + `eliminar/${idTituloAcademico}`);
  }
}

File: Backend/TrackrAPI/TrackrAPI/Services/Catalogo/TituloAcademicoService.cs
using TrackrAPI.Dtos.Catalogo;
using TrackrAPI.Models;
using TrackrAPI.Repositorys.Catalogo;

namespace TrackrAPI.Services.Catalogo
{
    public class TituloAcademicoService
    {
        private ITituloAcademicoRepository tituloAcademicoRepository;
        private TituloAcademicoValidatorService tituloAcademicoValidatorService;

        public TituloAcademicoService(ITituloAcademicoRepository tituloAcademicoRepository,
            TituloAcademicoValidatorService tituloAcademicoValidatorService)
        {
            this.tituloAcademicoRepository = tituloAcademicoRepository;
            this.tituloAcademicoValidatorService = tituloAcademicoValidatorService;
        }

        public IEnumerable<TituloAcademicoSelectorDto> ConsultarTodosParaSelector()
        {
            return tituloAcademicoRepository.ConsultarTodosParaSelector();
        }

        public TituloAcademicoDto ConsultarDto(int idTituloAcademico)
        {
            var tituloAcademico = tituloAcademicoRepository.ConsultarDto(idTituloAcademico);
            tituloAcademicoValidatorService.ValidarExistencia(tituloAcademico);
            return tituloAcademico;
        }

        public IEnumerable<TituloAcademicoGridDto> ConsultarTodosParaGrid()
        {
            return tituloAcademicoRepository.ConsultarTodosParaGrid();
        }

        public void Agregar(TituloAcademico tituloAcademico)
        {
            tituloAcademicoValidatorService.ValidarAgregar(tituloAcademico);
            tituloAcademicoRepository.Agregar(tituloAcademico);
        }

        public void Editar(TituloAcademico tituloAcademico)
        {
            tituloAcademicoValidatorService.ValidarEditar(tituloAcademico);
            tituloAcademicoRepository.Editar(tituloAcademico);
        }

        public void Eliminar(int idTituloAcademico)
        {
            var tituloAcademico = tituloAcademicoRepository.Consultar(idTituloAcademico);
            tituloAcademicoValidatorService.ValidarEliminar(idTituloAcademico);
            tituloAcademicoRepository.Eliminar(tituloAcademico);
        }

    }
}
